wallet-new/src/Cardano/Wallet/Kernel/Submission.hs

```diff
@@ -0,0 +1,393 @@
+makeLenses ''WalletSubmissionState
+
+instance Buildable Scheduler where
+    build ss =
+        let elems = ss ^. ssScheduled . to IntMap.toList
+        in bprint (F.later mapBuilder) elems
+
+
+instance Arbitrary Schedule where
+    arbitrary = Schedule <$> choose (0, maxRetries)
+
+-- Generates a random schedule by picking a slot >= of the input one but
+-- within a 'slot + 10' range, as really generating schedulers which generates
+-- things too far away in the future is not very useful for testing, if not
+-- testing that a scheduler will never reschedule something which cannot be
+-- reached.
+genScheduler :: Pending -> Slot -> Gen Scheduler
+genScheduler pending lowerBound = do
+    let pendingTxs  = pending ^. pendingTransactions . fromDb . to M.toList
+    slots <- vectorOf (length pendingTxs) (choose (lowerBound, lowerBound + 10))
+    retries    <- vectorOf (length pendingTxs) (choose (0, maxRetries))
+    return $ Scheduler $ List.foldl' updateFn mempty (zip3 slots pendingTxs retries)
+    where
+        updateFn acc (slot, (txId, txAux), retries) =
+            let s = Scheduled (txId, txAux, Schedule retries)
+            in IntMap.insertWith mappend slot [s] acc
+
+instance Arbitrary WalletSubmissionState where
+    arbitrary = do
+        pending   <- genPending (Core.ProtocolMagic 0)
+        slot      <- fmap getPositive arbitrary
+        scheduler <- genScheduler pending slot
+        return $ WalletSubmissionState pending scheduler slot
+
+genWalletSubmission :: ResubmissionFunction m -> Gen (WalletSubmission m)
+genWalletSubmission rho =
+    WalletSubmission <$> pure rho <*> arbitrary
+
+
+--
+-- Public API, as written in the spec
+--
+
+newWalletSubmission :: ResubmissionFunction m -> WalletSubmission m
+newWalletSubmission resubmissionFunction = WalletSubmission {
+      _wsResubmissionFunction = resubmissionFunction
+    , _wsState = newEmptyState
+    }
+    where
+        newEmptyState :: WalletSubmissionState
+        newEmptyState = WalletSubmissionState {
+              _wssPendingSet  = emptyPending
+            , _wssCurrentSlot = 0
+            , _wssScheduler   = Scheduler IntMap.empty
+            }
+
+-- | Informs the 'WalletSubmission' layer about new 'Pending' transactions.
+addPending :: WalletSubmission m -> Pending -> WalletSubmission m
+addPending ws newPending =
+    let ws' = ws & over (wsState . wssPendingSet) (unionPending newPending)
+    in schedule ws' newPending
+
+-- | Removes the input 'Pending' from the local 'WalletSubmission' pending set.
+remPending :: WalletSubmission m -> Pending -> WalletSubmission m
+remPending ws updatedPending =
+    ws & over (wsState . wssPendingSet) (flip differencePending updatedPending)
+
+-- | A \"tick\" of the scheduler.
+-- Returns the set transactions which needs to be droppped by the system as
+-- they likely exceeded the submission count and they have no chance to be
+-- adopted in a block, or that have simply been successfully retrasmitted.
```

Comment:
Typo. Also, I don't understand the last part "that have simply been successfully retransmitted". What do you mean?